### PR TITLE
feat: gossiping the latest valset in P2P layer in case they get pruned

### PR DESCRIPTION
**Pull request by SweeXordious**
## Overview

Closes https://github.com/celestiaorg/orchestrator-relayer/issues/556

## Checklist

- [ ] New and updated code has appropriate documentation
- [ ] New and updated code has new and/or updated testing
- [ ] Required CI checks are passing
- [ ] Visual proof for any user facing features like CLI or documentation updates
- [ ] Linked issues closed with keywords

### CODE REVIEW

File: p2p/dht.go
Comment (SweeXordious) [inline p2p/dht.go:105]:
```suggestion
// PutDataCommitmentConfirm encodes a data commitment confirm then puts its value to the DHT.
```

File: p2p/dht.go
Comment (SweeXordious) [inline p2p/dht.go:135]:
```suggestion
// PutValsetConfirm encodes a valset confirm then puts its value to the DHT.
```

File: p2p/dht.go
Comment (SweeXordious) [inline p2p/dht.go:165]:
```suggestion
// PutLatestValset encodes a valset then puts its value to the DHT.
```

File: p2p/dht.go
Comment (SweeXordious) [inline p2p/dht.go:168]:
```suggestion
// Returns an error if it fails.
```

File: p2p/dht.go
Comment (SweeXordious) [inline p2p/dht.go:181]:
```suggestion
// GetLatestValset looks for the latest valset in the DHT.
```

File: p2p/dht.go
Comment (SweeXordious) [inline p2p/dht.go:183]:
```suggestion
// Returns an error if it fails.
```

File: p2p/errors.go
Comment (SweeXordious) [inline p2p/errors.go:18]:
```suggestion
	ErrNoValidValueFound               = errors.New("no valid dht confirm value found")
```

File: p2p/keys.go
Comment (SweeXordious) [inline p2p/keys.go:37]:
```suggestion
	return "/" + LatestValsetNamespace + "/latest"
```

File: p2p/validators.go
Comment (SweeXordious) [inline p2p/validators.go:15]:
```suggestion
// Validate runs stateless checks on the provided confirm key and value.
```

File: p2p/validators.go
Comment (SweeXordious) [inline p2p/validators.go:77]:
```suggestion
// Validate runs stateless checks on the provided valset key and value.
```

File: p2p/validators.go
Comment (SweeXordious) [inline p2p/validators.go:92]:
```suggestion
// Select selects a valid dht valset value from multiple ones.
```

File: p2p/validators.go
Comment (SweeXordious) [inline p2p/validators.go:94]:
```suggestion
// returns an error of no valid value is found.
```

File: p2p/validators.go
Comment (SweeXordious) [inline p2p/validators.go:117]:
```suggestion
// Select selects a valid dht confirm value from multiple ones.
```

File: p2p/validators.go
Comment (SweeXordious) [inline p2p/validators.go:118]:
```suggestion
// returns an error of no valid value is found.
```

File: p2p/validators.go
Comment (SweeXordious) [inline p2p/validators.go:135]:
```suggestion
// Validate runs stateless checks on the provided confirm key and value.
```

File: p2p/validators.go
Comment (SweeXordious) [inline p2p/validators.go:195]:
```suggestion
// Select selects a valid dht confirm value from multiple ones.
// returns an error of no valid value is found.
```

File: p2p/validators.go
Comment (SweeXordious) [inline p2p/validators.go:201]:
```suggestion
		// choose the first correct value
```

File: p2p/validators_test.go
Comment (SweeXordious) [inline p2p/validators_test.go:464]:
```suggestion
			name:    "valid key and value",
```

File: p2p/validators_test.go
Comment (SweeXordious) [inline p2p/validators_test.go:482]:
```suggestion
			name:    "invalid value",
```

File: p2p/validators_test.go
Comment (SweeXordious) [inline p2p/validators_test.go:512]:
```suggestion
			name:    "no value",
```

File: types/latest_valset.go
Comment (SweeXordious) [inline types/latest_valset.go:28]:
```suggestion
// IsEmptyValset takes a valset and checks if it is empty.
```

File: orchestrator/broadcaster.go
Comment (evan-forbes) [inline orchestrator/broadcaster.go:6]: the standard thing is naming this `qgbtypes` instead of types2

File: p2p/dht.go
Comment (evan-forbes) [inline p2p/dht.go:194]: iirc, you mentioned that the the main concern with this approach is that we aren't verifying the valdiator set that is gossipped. Would that be possible though (in a futuer PR ofc)? We should be able to query the current validator set, and from that we should be able to measure the diff in power between the one we recieve that this one afaiu.

does that make sense or am I missing something?

File: types/latest_valset_test.go
Comment (evan-forbes) [inline types/latest_valset_test.go:33]: minor thing, but assert expects the expected value as the second arg and the compared value in the third.
```suggestion
	assert.Equal(t, expectedJSON, string(jsonData))
```

**Comment by evan-forbes** [comment]
only some minor comments, but otherwise LGTM

File: orchestrator/broadcaster.go
Comment (SweeXordious) [reply orchestrator/broadcaster.go]: [545ccee](https://github.com/celestiaorg/orchestrator-relayer/pull/560/commits/545ccee787c8c2b2674975f2af895dd35c783f89)

File: types/latest_valset_test.go
Comment (SweeXordious) [reply types/latest_valset_test.go:33]: good catch :+1:

File: p2p/dht.go
Comment (SweeXordious) [reply p2p/dht.go]: yes, we could, but I didn't want to add that check because it will be checked implicitly when sending the update to Blobstream. So, no need to do it explicitly here

Comment (evan-forbes) [reply p2p/dht.go]: do you mind pointing out that check really quickly for my own understanding? thanks!

File: orchestrator/broadcaster.go
Comment (evan-forbes) [reply orchestrator/broadcaster.go]: definitely not blocking, just for future reference, `celestiatypes` is better, but imo we should be specific with our imports to improve readability. for instance, what happens if we need to import `"github.com/celestiaorg/celestia-app/x/blob/types"` or  `"github.com/celestiaorg/celestia-core/types"` as well? personally I'm not a fan of naming everything types, but the sdk and tendermint don't agree lol

**Comment by SweeXordious** [comment]
Actually, I am facing an issue with the serialization of date to json. If you check the CI, you will see the tests failing, however, they're not failing locally. But when I use a different environment, they still fail expecting different time. So, I am thinking of actually omitting the time during serialization since it's not needed to create the hash or calculate the threshold etc.

File: p2p/dht.go
Comment (SweeXordious) [reply p2p/dht.go]: It will be done here:

https://github.com/celestiaorg/blobstream-contracts/blob/72b11fbc5ebd10c11e518bfd8b3f3a806d343b59/src/Blobstream.sol#L269-L276

in the contract

**Comment by SweeXordious** [comment]
@evan-forbes I did [755ff25](https://github.com/celestiaorg/orchestrator-relayer/pull/560/commits/755ff2509961e9a5b8ea14704e8389f2e273dd8b) to fix the issue of serialization via creating a struct that doesn't have the `time` field which was troublesome.

The `time` field is only used for pruning in the state machine. However, it's not part of the `hash` nor the `threshold` calculations. So, it made sense to remove it instead of dealing with serialization issues.

What do you think?

File: p2p/dht.go
Comment (evan-forbes) [reply p2p/dht.go]: but is it possible to gossip and invalid value? also, isn't the issue for folks deploying a new contract? would that check still be hit?

File: orchestrator/broadcaster.go
Comment (evan-forbes) [reply orchestrator/broadcaster.go]: yeah I'm 100% down for adding CI to look at import names. For packages with types I think we could define a linter that is generic enough to be useful. Not sure about other package names that frequently require renaming

**Review by evan-forbes — DISMISSED** [review]
I'm still thinking about the extra gossiping check. I think we might need something like it (see that thread), but that shouldn't block this PR

**Comment by evan-forbes** [comment]
the getting rid of the time makes sense. 

in the past with serializing time that needed to be kept, I've often had to overwrite marshaling methods to fix that one annoying type.

**Comment by SweeXordious** [comment]
> I've often had to overwrite marshaling methods to fix that one annoying type.

Can you link me to some code that does this? I'm curious

File: orchestrator/broadcaster.go
Comment (SweeXordious) [reply orchestrator/broadcaster.go]: https://github.com/celestiaorg/orchestrator-relayer/issues/563

File: p2p/dht.go
Comment (SweeXordious) [reply p2p/dht.go]: yes, an invalid value can be gossiped, but will be rejected by the contract when relaying events.

For the deployment part, I didn't implement the ability to use the P2P valset to deploy because it's super risky and there is no way to trust it.

**Comment by SweeXordious** [comment]
Added the changes discussed async. Will wait until we test this using an archive node to merge.

**Review by evan-forbes — APPROVED** [review]

**Comment by SweeXordious** [comment]
I will merge this and test it once we have an archive node ready for testing. And if something is broken, we can fix it in a separate PR.

**Pull request merged**
